File: Data Structures & Algorithms/count-connected-components/submission-1.cpp
class Solution {
    vector<vector<int>> adj;
    vector<bool> visited;
     void dfs(int node,int n){
        visited[node]=true;
        for(int pre:adj[node]){
            if(!visited[pre]) dfs(pre,n);
        }
    }
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        adj.assign(n,vector<int>());
       for(auto& edge:edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
       }
       int res=0;   
      visited.assign(n,false);
       for(int i=0;i<n;i++){
         if(!visited[i]){
            dfs(i,n);
            res++;
            }  }  
       return res;
    }
};

File: Data Structures & Algorithms/valid-tree/submission-1.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges)!=n-1:
            return False
        mp={}
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        visited= set()
        stack=[]
        stack.append(0)
        while len(stack):
            node= stack[-1]
            stack.pop()
            if node not in visited:
                visited.add(node)
                for nei in adj[node]:
                    if nei not in visited:
                        stack.append(nei)
        return len(visited)==n

        